### PR TITLE
IHostingStartup topic

### LINKED ISSUE #4438: Document IHostStartup (host lightup feature)
See Add docs for IHostedService #3352

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #4438 

[Internal Review Topic](https://review.docs.microsoft.com/en-us/aspnet/core/hosting/ihostingstartup?branch=pr-en-us-4967)

### CODE REVIEW

**Comment by guardrex** [comment]
@Tratcher Alrighty ... now that I've got us on the correct PR ...

I'll cover the relevant bits that I find from the discussion in https://github.com/aspnet/Home/issues/2051 and ping u back.

**Comment by guardrex** [comment]
@Tratcher I added two sections: Discovery to cover troubleshooting/debugging and a bit on globally disabling the automatically loaded hosting startup assemblies. What else would help in these regards?

File: aspnetcore/hosting/ihostingstartup.md [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+```
+%UserProfile%\.dotnet\x64\additionalDeps\StartupDiagnostics\
+```
+
+For examples of how to set environment variables for various operating systems, see [Working with multiple environments](xref:fundamentals/environments).
+
+## Discover loaded hosting startup assemblies
+
+To discover hosting startup assemblies loaded by the app or libraries added to the app, read the value of the [HostingStartupAssembliesKey](/dotnet/api/microsoft.aspnetcore.hosting.webhostdefaults.hostingstartupassemblieskey). The sample app reads the key into a `string` array and displays the result in the app's Index page:
```
Comment (Tratcher) [inline aspnetcore/hosting/ihostingstartup.md]: First check the application logs:
https://github.com/aspnet/Hosting/blob/86df76440b450daa3972d5b317026b58f718b803/src/Microsoft.AspNetCore.Hosting/Internal/WebHost.cs#L136-L151

File: aspnetcore/hosting/ihostingstartup.md [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+## Disable automatic loading of hosting startup assemblies
+
+There are two ways to disable the automatic loading of hosting startup assemblies:
+
+* Set the [Prevent Hosting Startup](xref:fundamentals/hosting#prevent-hosting-startup) host configuration setting.
+* Set the `ASPNETCORE_preventHostingStartup` environment variable.
+
+When either the host setting or the environment variable is set to `true` or `1`, hosting startup assemblies aren't automatically loaded. If both are set, the host setting controls the behavior.
+
+It isn't currently possible to selectively disable a hosting startup assembly added by a library unless the library offers its own configuration option. Disabling hosting startup assemblies using the host setting or environment variable disables them globally and may disable several features of an app.
```
Comment (Tratcher) [inline aspnetcore/hosting/ihostingstartup.md]: want to link currently to the 2.1 bug that added this feature?
https://github.com/aspnet/Hosting/pull/1243

Comment (guardrex) [reply aspnetcore/hosting/ihostingstartup.md]: We don't generally discuss 2.1 features in the 2.0-era docs. @Rick-Anderson Do you want to make an exception here? I can say something to the effect that 'a future release will offer this feature' and then link that issue. I believe we have done that a few times in the past.

**Comment by guardrex** [comment]
> move these new sections to the very top right after the opening paragraph

:+1:

File: aspnetcore/hosting/ihostingstartup.md (hunk 2 above)
Comment (Rick-Anderson) [reply aspnetcore/hosting/ihostingstartup.md]: @guardrex  yes

**Comment by guardrex** [comment]
@Tratcher :+1: See how those updates look.

[Internal Review Topic](https://review.docs.microsoft.com/en-us/aspnet/core/hosting/ihostingstartup?branch=pr-en-us-4967)

File: aspnetcore/hosting/ihostingstartup.md [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+When building an `IHostingStartup` project, the dependencies file (*\*.deps.json*) sets the `runtime` location of the assembly to the *bin* folder:
+
+[!code-json[Main](ihostingstartup/snapshot_sample/StartupFeature1.deps.json?range=2-13&highlight=8)]
+
+Only part of the file is shown. The assembly name in the example is `StartupFeature`.
+
+### Update the dependencies file
+
+In order for the feature to activate for an app, the `runtime` location in the *\*.deps.json* file must be changed to reflect the location of the feature's runtime assembly in the runtime store. Prefix the `runtime` location with `lib/netcoreapp2.0/`:
```
Comment (Rick-Anderson) [inline aspnetcore/hosting/ihostingstartup.md]: Way too long. Alt:
The `runtime` location for features is specified in the *\*.deps.json* file. To active a feature,  the `runtime` must include the location of the feature's runtime assembly. Prefix the `runtime` location with `lib/netcoreapp2.0/`:

Or can you make a bullet list?

File: aspnetcore/hosting/ihostingstartup.md [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+```
+%UserProfile%\.dotnet\x64\additionalDeps\StartupDiagnostics\
+```
+
+For examples of how to set environment variables for various operating systems, see [Working with multiple environments](xref:fundamentals/environments).
+
+## Sample app
+
+The [sample app](https://github.com/aspnet/Docs/tree/master/aspnetcore/hosting/ihostingstartup/sample/) ([how to download](xref:tutorials/index#how-to-download-a-sample)) uses `IHostingStartup` to create a diagnostics tool. The tool adds two middlewares to the app at startup that provide diagnostic information:
```
Comment (Rick-Anderson) [inline aspnetcore/hosting/ihostingstartup.md]: provides ?

File: aspnetcore/hosting/ihostingstartup.md [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+By [Luke Latham](https://github.com/guardrex)
+
+An [IHostingStartup](/dotnet/api/microsoft.aspnetcore.hosting.ihostingstartup) implementation allows adding features to an app at startup from outside of the app's `Startup` class. For example, an external tooling library can use an `IHostingStartup` implementation to provide additional configuration providers or services to an app. `IHostingStartup` *is available in ASP.NET Core 2.0 and later.*
+
+[View or download sample code](https://github.com/aspnet/Docs/tree/master/aspnetcore/hosting/ihostingstartup/sample/) ([how to download](xref:tutorials/index#how-to-download-a-sample))
+
+## Discover loaded hosting startup assemblies
+
+To discover hosting startup assemblies loaded by the app or by libraries, enable logging and check the application logs. Errors that occur when loading assemblies are also logged. Loaded hosting startup assemblies are logged at the Debug level, and errors are always logged.
```
Comment (Rick-Anderson) [inline aspnetcore/hosting/ihostingstartup.md]: Errors that occur when loading assemblies are ~also~ logged

Comment (Rick-Anderson) [reply aspnetcore/hosting/ihostingstartup.md]: >and errors are always logged.

How about
and all errors are logged.

File: aspnetcore/hosting/ihostingstartup.md [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+It's also possible to read the value of the [HostingStartupAssembliesKey](/dotnet/api/microsoft.aspnetcore.hosting.webhostdefaults.hostingstartupassemblieskey). The sample app reads the key into a `string` array and displays the result in the app's Index page:
```
Comment (Rick-Anderson) [inline aspnetcore/hosting/ihostingstartup.md]: >It's also possible to read the value of the
You can read the value of the

Comment (Rick-Anderson) [reply aspnetcore/hosting/ihostingstartup.md]: Or better yet, nuke that entire sentence. The following sample reads the HostingStartupAssembliesKey key into ...

File: aspnetcore/hosting/ihostingstartup.md (hunk 3 above)
Comment (guardrex) [reply aspnetcore/hosting/ihostingstartup.md]: I'll try a shorter version of this that breaks the sentence as you suggest.

File: aspnetcore/hosting/ihostingstartup.md (hunk 4 above)
Comment (guardrex) [reply aspnetcore/hosting/ihostingstartup.md]: The "two middlewares" "provide" (plural verb) 'the feature.' I think we're ok here.

File: aspnetcore/hosting/ihostingstartup.md (hunk 6 above)
Comment (guardrex) [reply aspnetcore/hosting/ihostingstartup.md]: I'll make the change, but this is an alternative to the first option of looking at the logs. This loses it's conversational tone by jumping right to 'the sample does Y for this' here.

File: aspnetcore/hosting/ihostingstartup.md (hunk 4 above)
Comment (guardrex) [reply aspnetcore/hosting/ihostingstartup.md]: I guess the reason that I'm focusing on "two middlewares" and not "the tool" is that the object of "that" feels like it refers to the "two middlewares" more than "the tool." This is the way that I would speak the sentence verbally.

Comment (Rick-Anderson) [reply aspnetcore/hosting/ihostingstartup.md]: I agree. That's why I put ?

Comment (guardrex) [reply aspnetcore/hosting/ihostingstartup.md]: lol :smile: I don't take anything you suggest lightly!

**Pull request merged**